patroni/ha.py

```diff
@@ -304,19 +347,45 @@ def is_healthiest_node(self):
         members = {m.name: m for m in self.cluster.members + self.old_cluster.members}
         return self._is_healthiest_node(members.values())
 
-    def demote(self, delete_leader=True):
-        if delete_leader:
-            self.state_handler.stop()
+    def release_leader_key_voluntarily(self):
+        self.dcs.delete_leader()
+        self.touch_member()
+        self.dcs.reset_cluster()
+        logger.info("Leader key released")
+
+    def demote(self, mode):
```

Comment:
https://github.com/zalando/patroni/pull/316/commits/e538e322bcc530b55b407c2c0d150049e21a519b commit has changed slightly `process_healthy_cluster` behavior. Now it calls `self.demote(delete_leader=False)` immediately after it failed to update the leader lock.

Please update from master and don't forget to fix `process_healthy_cluster`. I think unit tests should catch.